src/librustc_mir/interpret/validity.rs

```diff
@@ -502,12 +505,14 @@ impl<'rt, 'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>>
                     _ => false,
                 }
             } => {
-                let mplace = if op.layout.is_zst() {
+                let mplace = match *op {
                     // it's a ZST, the memory content cannot matter
-                    MPlaceTy::dangling(op.layout, self.ecx)
-                } else {
-                    // non-ZST array/slice/str cannot be immediate
-                    op.to_mem_place()
+                    Operand::Immediate(_) if op.layout.is_zst() =>
+                        // invent an aligned mplace
+                        MPlaceTy::dangling(op.layout, self.ecx),
+                    // FIXME: what about single element arrays? They can be Scalar layout I think
+                    Operand::Immediate(_) => bug!("non-ZST array/slice cannot be immediate"),
+                    Operand::Indirect(_) => op.to_mem_place(),
```

Comment:
What is this about? I do not quite like inspecting the `Operand` here, if we have to do that I feel that means something is wrong.

Reply:
I could just call `to_mem_place`, but the `bug` message can be more concrete here.

The reason I'm not doing the old check anymore is that in the case of `Indirect` operands, we'd erase the pointer in favour of a definitely aligned one. This makes the alignment check useless.

I can keep the original logic, but that would turn the zst-arm into `return Ok(());` (which was the original logic minus the length extraction.)

Reply:
You mean alignment check of the operand itself? That shouldn't be necessary, we have an invariant that we only call validation on aligned and allocated operands. When we recursively check a reference, we check these things before pushing it to the `todo`.

And no, the ZST arm can *not* be `return Ok(())` (and there was a comment somewhere saying that): `[!; 1]` is a ZST and better be checked.

Reply:
The code here is specialized for `ty::Int(..) | ty::Uint(..) | ty::Float(..) => true,` elements of arrays/slices. So if the type is zst (an empty array) we can just return. If the type is a slice, we can early return if the length is zero. In all other cases, the pointer cannot be integral.